elytron/ELY-2078-encryption-for-FilesystemSecurityRealm.adoc

```diff
@@ -0,0 +1,250 @@
+- The storage of the principal and passwords are handled separately. The principal is stored as the filename and encoded in base32, whereas the password is stored within the file and encrypted with a secret key and encoded with base64.
+- At the moment the only encoding right now for the principal and password (if plaintext is the choice of storage) is a base64 encoding, but this is very easily decodeable, very quickly. The solution is to encrypt it with a SecretKey, and then encode it wih Base64 afterwards.
+- Due to design limitations, the principal/username cannot contain illegal file characters, it requires the same output for every input, and it must be possible to determine the username from the filename. Due to these requirements, hashing would not work. Along with that the CipherUtil SecretKey methods would also not work as it uses a sort of randomization that ensures not every identical input will have the same encrypted output. As a result it was decided to just encode the principal with a base32 encoding and just encrypt the remaining information.
+
+- There were 2 options considered for the SecretKey's. The first was to encrypt each parameter: password, roles/attributes, with a separate SecretKey each, and store all the different SecretKey's in one shared `PropertiesCredentialStore`. The other option was to encrypt the different parameters with the same SecretKey and store it in one `PropertiesCredentialStore`. The decision chosen was to use just one SecretKey and the reasoning for that was because if there was a breach in security and the password for the PropertiesCredentialStore was leaked, if all the SecretKey's were in one `PropertiesCredentialStore`, then getting access to 3 SecretKey's as opposed to 1 SecretKey is no different and has no advantages.
+
+
+- The execution of creating an encrypted FileSystem realm would be as follows
+    * Create a CredentialStore (used to store the SecretKey). This would create a credential store and store a secret key inside it with an alias "key"
+    * Then the user would create a FileSystemRealm and pass in the reference of the credential store as well as passing in the alias of the secret key defined in the Elytron subsystem.
+- It will be necessary to modify the `FileSystemRealmDefinition` to incorporate the option to add a path to the credential store as well as the alias for the SecretKey if the user want's to encrypt their FileSystemRealm. This would go under the `RealmAddHandler` class in the `performRuntime()` method. Also will need to add a `SimpleAttributeDefinition` to pass in a SecretKey into the FileSystemSecurityRealm builder. Need to also update the `ElytronDescriptionConstants` with the constants containing encryption information too.
+
+=== Filesystem Integrity
+In the FileSystemSecurityRealm, a method will be added to generate the signature for any given identity file. Another method will be added to verify the signature of a given file.
```

Comment:
Would be good to mention that the integrity changes are part of planned upcoming changes but aren't targetting 26.1.

Reply:
Removed